compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -2072,6 +2116,49 @@ fn determine_capture_info(
     }
 }
 
+/// Truncates `place` to have up to `len` projections.
+/// `curr_mode` is the current required capture kind for the place.
+/// Returns the truncated `place` and the updated required capture kind.
+///
+/// Note: Capture kind changes from `MutBorrow` to `UniqueImmBorrow` if the truncated part of the `place`
+/// contained `Deref` of `&mut`.
+fn truncate_place_to_len(
+    mut place: Place<'tcx>,
+    curr_mode: ty::UpvarCapture<'tcx>,
```

Comment:
I wonder if we want this to be

```rust
truncate_place_to_len(mut place: Place<'tcx>, curr_mode: &mut ty::UpvarCapture<'tcx>, len: usize)
```

or even

```rust
truncate_place_to_len(place: &mut Place<'tcx>, curr_mode: &mut ty::UpvarCapture<'tcx>, len: usize)
```

It seemed like a lot of the callers (*almost* all) were doing

```rust
let (..., mode) = truncate_place_to_len(place, curr_mode, ...);
curr_mode = mode;
```

which is a bit silly.

Reply:
My thoughts:
- Looking at just the function all it's not directly evident that the `mode` has changed, might need a better name.
- Some codepaths don't really have a `&mut Place` (eg:  within Delegate -- we need to handle repr packed) so that might need adding a line to create an owned `Place` variable.

Can we use unstable features? https://github.com/rust-lang/rust/issues/71126 looks is near stabilization and the syntax is what I'd prefer, i.e. directly update the variables than destructure and then assign

Reply:
We are permitted to use unstable features, yes.